src/libmongoc/src/mongoc/mongoc-ts-pool.c

```diff
@@ -0,0 +1,282 @@
+#include "mongoc-ts-pool-private.h"
+#include "common-thread-private.h"
+
+#include "bson/bson.h"
+
+#if defined(_MSC_VER)
+typedef double _max_align_type;
+#elif defined(__APPLE__)
+typedef long double _max_align_type;
+#else
+/* Based on the GCC max_align_t definition */
+typedef struct _max_align_type {
+   long long maxalign_1 __attribute__ ((__aligned__ (__alignof__(long long))));
+   long double maxalign_2
+      __attribute__ ((__aligned__ (__alignof__(long double))));
+} _max_align_type;
+#endif
+
+/**
+ * Toggle this to enable/disable checks that all items are returned to the pool
+ * before the pool is destroyed
+ */
+enum { AUDIT_POOL_ENABLED = 0 };
+
+typedef struct pool_node {
+   struct pool_node *next;
+   mongoc_ts_pool *owner_pool;
+
+   _max_align_type data[];
+} pool_node;
+
+struct mongoc_ts_pool {
+   mongoc_ts_pool_params params;
+   pool_node *head;
+   /* Number of elements in the pool */
+   int32_t size;
+   bson_mutex_t mtx;
+   /* Number of elements that the pool has given to users.
+    * If AUTO_POOL_ENABLED is zero, this member is unused
+    */
+   int32_t outstanding_items;
+};
+
+/**
+ * @brief Check whether we should drop the given node from the pool
+ */
+static bool
+_should_prune (const pool_node *node)
+{
+   mongoc_ts_pool *pool = node->owner_pool;
+   return pool->params.prune_predicate &&
+          pool->params.prune_predicate (node->data, pool->params.userdata);
+}
+
+/**
+ * @brief Create a new pool node and contained element.
+ *
+ * @return pool_node* A pointer to a constructed element, or NULL if the
+ * constructor sets `error`
+ */
+static pool_node *
+_new_item (mongoc_ts_pool *pool, bson_error_t *error)
+{
+   pool_node *node =
+      bson_malloc0 (sizeof (pool_node) + pool->params.element_size);
```

Comment:
I want to confirm: is this allocation size correct given the flexible data member in `pool_node` does not contribute to `sizeof (pool_node)` (does `pool->params.element_size` correctly account for possible access of `node->data`)?

Reply:
That is correct, yes.